File: packages/go/models/access_rule.go
package models

import (
	"fmt"
	"strconv"
)

// RuleAddress is a source or destination address for an access rule.
type RuleAddress struct {
	Any   bool
	Name  string
	Group string
}

// RuleService is a service matcher for an access rule.
type RuleService struct {
	Any  bool
	Name string
}

// RulePriority controls SonicOS rule ordering.
type RulePriority struct {
	Auto  bool
	Value *int
}

// AccessRule is an IPv4 access rule.
type AccessRule struct {
	Name                string
	FromZone            string
	ToZone              string
	Action              string // allow, deny, discard
	Enabled             bool
	Log                 bool
	Priority            RulePriority
	SourceAddress       RuleAddress
	DestinationAddress  RuleAddress
	Service             RuleService
	Comment             string
}

func stringFromAny(v any) string {
	if v == nil {
		return ""
	}
	switch t := v.(type) {
	case string:
		return t
	case float64:
		return strconv.FormatInt(int64(t), 10)
	case bool:
		return strconv.FormatBool(t)
	default:
		return fmt.Sprint(t)
	}
}

func boolFromAny(v any, defaultVal bool) bool {
	if v == nil {
		return defaultVal
	}
	b, ok := v.(bool)
	if !ok {
		return defaultVal
	}
	return b
}

// AccessRuleToAPIEnvelope returns {"access_rule":{"ipv4":{...}}}.
func AccessRuleToAPIEnvelope(r *AccessRule) map[string]any {
	inner := map[string]any{
		"from":        r.FromZone,
		"to":          r.ToZone,
		"action":      r.Action,
		"enabled":     r.Enabled,
		"log":         r.Log,
		"source":      ruleAddrWire(r.SourceAddress),
		"destination": ruleAddrWire(r.DestinationAddress),
		"service":     ruleSvcWire(r.Service),
	}
	if r.Name != "" {
		inner["name"] = r.Name
	}
	if r.Priority.Auto {
		inner["priority"] = map[string]any{"auto": true}
	} else if r.Priority.Value != nil {
		inner["priority"] = map[string]any{"value": *r.Priority.Value}
	}
	if r.Comment != "" {
		inner["comment"] = r.Comment
	}
	return map[string]any{"access_rule": map[string]any{"ipv4": inner}}
}

func ruleAddrWire(a RuleAddress) map[string]any {
	if a.Any {
		return map[string]any{"address": map[string]any{"any": true}}
	}
	if a.Name != "" {
		return map[string]any{"address": map[string]any{"name": a.Name}}
	}
	if a.Group != "" {
		return map[string]any{"address": map[string]any{"group": a.Group}}
	}
	return map[string]any{"address": map[string]any{"any": true}}
}

func ruleSvcWire(s RuleService) map[string]any {
	if s.Any {
		return map[string]any{"any": true}
	}
	if s.Name != "" {
		return map[string]any{"name": s.Name}
	}
	return map[string]any{"any": true}
}

// AccessRuleFromAPIItem parses a list item or singular response into AccessRule.
func AccessRuleFromAPIItem(data map[string]any) (*AccessRule, error) {
	inner := data
	if ar, ok := data["access_rule"].(map[string]any); ok {
		inner = ar
	}
	if ipv4, ok := inner["ipv4"].(map[string]any); ok {
		inner = ipv4
	}

	r := &AccessRule{
		Name:     stringFromAny(inner["name"]),
		FromZone: stringFromAny(inner["from"]),
		ToZone:   stringFromAny(inner["to"]),
		Action:   stringFromAny(inner["action"]),
		Enabled:  boolFromAny(inner["enabled"], true),
		Log:      inner["log"] == true,
		Priority: RulePriority{Auto: true},
	}

	if p, ok := inner["priority"].(map[string]any); ok {
		if p["auto"] == true {
			r.Priority = RulePriority{Auto: true}
		} else if v, ok := p["value"].(float64); ok {
			vi := int(v)
			r.Priority = RulePriority{Auto: false, Value: &vi}
		}
	}

	if src, ok := inner["source"].(map[string]any); ok {
		if addr, ok := src["address"].(map[string]any); ok {
			r.SourceAddress = parseRuleAddr(addr)
		}
	}
	if dst, ok := inner["destination"].(map[string]any); ok {
		if addr, ok := dst["address"].(map[string]any); ok {
			r.DestinationAddress = parseRuleAddr(addr)
		}
	}
	if svc, ok := inner["service"].(map[string]any); ok {
		r.Service = parseRuleSvc(svc)
	}
	if r.Action == "" {
		r.Action = "allow"
	}
	r.Comment = stringFromAny(inner["comment"])
	return r, nil
}

func parseRuleAddr(addr map[string]any) RuleAddress {
	if addr["any"] == true {
		return RuleAddress{Any: true}
	}
	if n, ok := addr["name"].(string); ok && n != "" {
		return RuleAddress{Name: n}
	}
	if g, ok := addr["group"].(string); ok && g != "" {
		return RuleAddress{Group: g}
	}
	return RuleAddress{Any: true}
}

func parseRuleSvc(svc map[string]any) RuleService {
	if svc["any"] == true {
		return RuleService{Any: true}
	}
	if n, ok := svc["name"].(string); ok && n != "" {
		return RuleService{Name: n}
	}
	return RuleService{Any: true}
}

File: packages/typescript/src/models/accessRule.ts
/**
 * SonicOS IPv4 access rule.
 */

import { z } from "zod";

export const AccessRuleActionSchema = z.enum(["allow", "deny", "discard"]);
export type AccessRuleAction = z.infer<typeof AccessRuleActionSchema>;

export const RuleAddressSchema = z.object({
  any: z.boolean().default(false),
  name: z.string().optional(),
  group: z.string().optional(),
});

export const RuleServiceSchema = z.object({
  any: z.boolean().default(false),
  name: z.string().optional(),
});

export const RulePrioritySchema = z.object({
  auto: z.boolean().default(true),
  value: z.number().int().optional(),
});

export const AccessRuleSchema = z.object({
  name: z.string().max(31).optional(),
  fromZone: z.string(),
  toZone: z.string(),
  action: AccessRuleActionSchema.default("allow"),
  enabled: z.boolean().default(true),
  log: z.boolean().default(false),
  priority: RulePrioritySchema.default({ auto: true }),
  sourceAddress: RuleAddressSchema.default({ any: true }),
  destinationAddress: RuleAddressSchema.default({ any: true }),
  service: RuleServiceSchema.default({ any: true }),
  comment: z.string().max(255).optional(),
});

export type RuleAddress = z.infer<typeof RuleAddressSchema>;
export type RuleService = z.infer<typeof RuleServiceSchema>;
export type RulePriority = z.infer<typeof RulePrioritySchema>;
export type AccessRule = z.infer<typeof AccessRuleSchema>;

function addrWire(addr: RuleAddress): Record<string, unknown> {
  if (addr.any) return { address: { any: true } };
  if (addr.name) return { address: { name: addr.name } };
  if (addr.group) return { address: { group: addr.group } };
  return { address: { any: true } };
}

function svcWire(svc: RuleService): Record<string, unknown> {
  if (svc.any) return { any: true };
  if (svc.name) return { name: svc.name };
  return { any: true };
}

export function accessRuleToApiDict(rule: AccessRule): Record<string, unknown> {
  const inner: Record<string, unknown> = {
    from: rule.fromZone,
    to: rule.toZone,
    action: rule.action,
    enabled: rule.enabled,
    log: rule.log,
    source: addrWire(rule.sourceAddress),
    destination: addrWire(rule.destinationAddress),
    service: svcWire(rule.service),
  };
  if (rule.name) inner["name"] = rule.name;
  if (rule.priority.auto) {
    inner["priority"] = { auto: true };
  } else if (rule.priority.value != null) {
    inner["priority"] = { value: rule.priority.value };
  }
  if (rule.comment) inner["comment"] = rule.comment;
  return { access_rule: { ipv4: inner } };
}

export function accessRuleFromApiResponse(data: Record<string, unknown>): AccessRule {
  let inner = data;
  if ("access_rule" in inner && typeof inner["access_rule"] === "object" && inner["access_rule"]) {
    inner = inner["access_rule"] as Record<string, unknown>;
  }
  if ("ipv4" in inner && typeof inner["ipv4"] === "object" && inner["ipv4"]) {
    inner = inner["ipv4"] as Record<string, unknown>;
  }

  const prioRaw = inner["priority"] as Record<string, unknown> | undefined;
  let priority: RulePriority = { auto: true };
  if (prioRaw && typeof prioRaw === "object") {
    if (prioRaw["auto"] === true) priority = { auto: true };
    else if (typeof prioRaw["value"] === "number") priority = { auto: false, value: prioRaw["value"] };
  }

  const src = (inner["source"] as Record<string, unknown>)?.["address"] as Record<string, unknown> | undefined;
  let sourceAddress: RuleAddress = { any: true };
  if (src?.["any"]) sourceAddress = { any: true };
  else if (typeof src?.["name"] === "string") sourceAddress = { any: false, name: src["name"] };
  else if (typeof src?.["group"] === "string") sourceAddress = { any: false, group: src["group"] };

  const dst = (inner["destination"] as Record<string, unknown>)?.["address"] as
    | Record<string, unknown>
    | undefined;
  let destinationAddress: RuleAddress = { any: true };
  if (dst?.["any"]) destinationAddress = { any: true };
  else if (typeof dst?.["name"] === "string") destinationAddress = { any: false, name: dst["name"] };
  else if (typeof dst?.["group"] === "string") destinationAddress = { any: false, group: dst["group"] };

  const svc = inner["service"] as Record<string, unknown> | undefined;
  let service: RuleService = { any: true };
  if (svc && typeof svc === "object") {
    if (svc["any"]) service = { any: true };
    else if (typeof svc["name"] === "string") service = { any: false, name: svc["name"] };
  }

  return AccessRuleSchema.parse({
    name: inner["name"] != null ? String(inner["name"]) : undefined,
    fromZone: String(inner["from"] ?? ""),
    toZone: String(inner["to"] ?? ""),
    action: (inner["action"] as AccessRuleAction) ?? "allow",
    enabled: inner["enabled"] !== false,
    log: inner["log"] === true,
    priority,
    sourceAddress,
    destinationAddress,
    service,
    comment: inner["comment"] != null ? String(inner["comment"]) : undefined,
  });
}
